pallets/merkle/src/utils/hasher.rs

```diff
@@ -74,59 +109,93 @@ impl Setup {
 					let sr = Scalar::from_bytes_mod_order(slice_to_bytes_32(xr));
-					Poseidon_hash_2(sl, sr, &DEFAULT_POSEIDON_HASHER).to_bytes().to_vec()
+					Ok(Poseidon_hash_2(sl, sr, &DEFAULT_POSEIDON_HASHER).to_bytes().to_vec())
 				}
-				_ => panic!("Unsupported hash function"),
+				_ => Err(Error::<C>::Unimplemented),
 			},
-			_ => panic!("Unsupported backend"),
+			Backend::Arkworks(Curve::Bls381, _) => match self.hasher {
+				HashFunction::PoseidonDefault => {
+					let mut bytes = Vec::new();
+					bytes.extend(xl);
+					bytes.extend(xr);
+					let res = PoseidonCRH3::evaluate(&POSEIDON_PARAMETERS, &bytes).unwrap();
+					let bytes_res = to_bytes![res];
+					let bytes = match bytes_res {
+						Ok(bytes) => bytes,
+						Err(_) => return Err(Error::<C>::HashingFailed),
+					};
+					Ok(bytes)
+				}
+				_ => Err(Error::<C>::Unimplemented),
+			},
+			_ => Err(Error::<C>::Unimplemented),
 		}
 	}
 
-	pub fn get_bulletproofs_poseidon(&self) -> &Poseidon {
-		match self.backend {
-			Backend::Bulletproofs => match self.hasher {
-				HashFunction::PoseidonDefault => &DEFAULT_POSEIDON_HASHER,
-				_ => panic!("Hasher is not default poseidon hasher"),
-			},
-			_ => panic!("Backend not bulletproofs"),
+	pub fn get_default_bulletproofs_poseidon<C: Config>(&self) -> Result<&Poseidon, Error<C>> {
+		match (&self.backend, &self.hasher) {
+			(Backend::Bulletproofs(Curve::Curve25519), HashFunction::PoseidonDefault) => Ok(&DEFAULT_POSEIDON_HASHER),
+			_ => Err(Error::<C>::Unimplemented),
 		}
 	}
 
-	pub fn generate_zero_tree(&self, depth: usize) -> (Vec<ScalarBytes>, ScalarBytes) {
-		let zero_tree = match self.hasher {
-			HashFunction::PoseidonDefault => {
-				gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox)
-			}
-			_ => gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox),
-		};
-		(
-			zero_tree[0..depth].iter().map(|x| x.to_vec()).collect(),
-			zero_tree[depth].to_vec(),
-		)
+	pub fn get_default_proving_key<C: Config>(&self) -> Result<&ProvingKey<Bls12_381>, Error<C>> {
+		match (&self.backend, &self.hasher) {
+			(Backend::Arkworks(Curve::Bls381, Snark::Groth16), _) => Ok(&PROVING_KEY),
+			_ => Err(Error::<C>::Unimplemented),
+		}
+	}
+
+	pub fn generate_zero_tree<C: Config>(&self, depth: usize) -> Result<(Vec<ScalarBytes>, ScalarBytes), Error<C>> {
+		match self.backend {
+			Backend::Bulletproofs(Curve::Curve25519) => match self.hasher {
+				HashFunction::PoseidonDefault => {
+					let zero_tree = gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox);
+					Ok((
+						zero_tree[0..depth].iter().map(|x| x.to_vec()).collect(),
+						zero_tree[depth].to_vec(),
+					))
+				}
+				_ => Err(Error::<C>::Unimplemented),
+			},
+			Backend::Arkworks(Curve::Bls381, _) => match self.hasher {
+				HashFunction::PoseidonDefault => {
+					let res = gen_empty_hashes::<TreeConfig>(&(), &POSEIDON_PARAMETERS)
```

Comment:
We should consider adding a more verbose name for ARKWORKS_POSEIDON_PARAMETERS and BULLETPROOF_POSEIDON_PARAMETERS.

Or more concretely, since the parameters are likely curve specific. BLS381_POSEIDON_PARAMETERS. Thoughts?

Reply:
I went with `DEFAULT_BULLETPROOFS_POSEIDON_HASHER` and `DEFAULT_ARKWORKS_POSEIDON_PARAMETERS` which will use the default hardcoded curve. For future ones, we can add the name of the curve.